tokio/src/sync/watch.rs

```diff
@@ -610,6 +610,55 @@ impl<T> Receiver<T> {
         }
     }
 
+    /// This function is similar to [`Receiver::changed()`], but it takes a closure that is
+    /// called with a reference to the new value. If the closure returns `true`,
+    /// then the function returns immediately. Otherwise, it waits for a new
+    /// value and calls the closure again.
+    ///
+    /// [`Receiver::changed()`]: crate::sync::watch::Receiver::changed
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::watch;
+    ///
+    /// #[tokio::main]
+    ///
+    /// async fn main() {
+    ///     let (tx, _rx) = watch::channel("hello");
+    ///
+    ///     tx.send("goodbye").unwrap();
+    ///     
+    ///     // here we subscribe to a second receiver
+    ///     // now in case of using `changed` we would have
+    ///     // to first check the current value and then wait
+    ///     // for changes or else `changed` would hang.
+    ///     let mut rx2 = tx.subscribe();
+    ///     
+    ///     // in place of changed we have use `wait_for`
+    ///     // which would automatically check the current value
+    ///     // and wait for changes until the closure returns true.
+    ///     assert!(rx2.wait_for(|val| *val == "goodbye").await.is_ok());
+    ///     assert_eq!(*rx2.borrow(), "goodbye");
+    /// }
+    /// ```
+
+    pub async fn wait_for(
+        &mut self,
+        mut f: impl FnMut(&T) -> bool,
+    ) -> Result<bool, error::RecvError> {
```

Comment:
You never return an error, so why a `Result`?

Reply:
Oh so sorry oversight, I meant to return errors, would it be better to just return booleans? or should we return errors?

Reply:
I think the correct way to propagate the error would be like this? instead of checking like this
```diff
diff --git a/tokio/src/sync/watch.rs b/tokio/src/sync/watch.rs
index df2674dd..309eea38 100644
--- a/tokio/src/sync/watch.rs
+++ b/tokio/src/sync/watch.rs
@@ -653,10 +653,7 @@ impl<T> Receiver<T> {
                 return Ok(true);
             }
 
-            let changed = self.changed().await;
-            if changed.is_err() {
-                return Ok(false);
-            }
+            self.changed().await?;
         }
     }
 
```

Reply:
Updated in [46196fb](https://github.com/tokio-rs/tokio/pull/5611/commits/46196fb73ea4c5f78fd75e8472d2ef1b3cc89ef5) does it look ok?